prov/rxm/src/rxm_cq.c

```diff
@@ -708,6 +708,82 @@ static inline void rxm_cq_repost_rx_buffers(struct rxm_ep *rxm_ep)
 	}
 }
 
+static int rxm_cq_reprocess_directed_recvs(struct rxm_recv_queue *recv_queue)
+{
+	struct rxm_rx_buf *rx_buf;
+	struct dlist_entry *entry, *tmp_entry;
+	struct rxm_recv_match_attr match_attr;
+	struct dlist_entry rx_buf_list;
+	int ret = 0, count = 0;
+
+	dlist_init(&rx_buf_list);
+
+	recv_queue->rxm_ep->res_fastlock_acquire(&recv_queue->lock);
+
+	dlist_foreach_container_safe(&recv_queue->unexp_msg_list,
+				     struct rxm_rx_buf, rx_buf,
+				     unexp_msg.entry, tmp_entry) {
+		if (rx_buf->unexp_msg.addr == rx_buf->conn->handle.fi_addr)
+			continue;
+
+		assert(rx_buf->unexp_msg.addr == FI_ADDR_NOTAVAIL);
+
+		match_attr.addr = rx_buf->unexp_msg.addr =
+			rx_buf->conn->handle.fi_addr;
+		match_attr.tag = rx_buf->unexp_msg.tag;
+
+		entry = dlist_remove_first_match(&recv_queue->recv_list,
+						 recv_queue->match_recv,
+						 &match_attr);
+		if (!entry)
+			continue;
+
+		dlist_remove(&rx_buf->unexp_msg.entry);
+		rx_buf->recv_entry = container_of(entry, struct rxm_recv_entry,
+						  entry);
+		dlist_insert_tail(&rx_buf->unexp_msg.entry, &rx_buf_list);
+	}
+	recv_queue->rxm_ep->res_fastlock_release(&recv_queue->lock);
+
+	while (!dlist_empty(&rx_buf_list)) {
+		dlist_pop_front(&rx_buf_list, struct rxm_rx_buf,
+				rx_buf, unexp_msg.entry);
+		ret = rxm_cq_handle_rx_buf(rx_buf);
+		if (ret)
+			return ret;
```

Comment:
I don't think we can return here without risk of leaking unexp_msg's.  I don't know how to handle errors in this case.  :/

Reply:
Error from rxm_cq_handle_rx_buf would probably be unrecoverable. If we make the assumption that the app would halt, I think we're fine here. Otherwise, if we want to put them back in the unexpected queue, we may have to create some sort of place holders in the queue for these removed messages.

Reply:
Can we write a failed entry into the CQ, cleanup the buffer, and just move on?

Reply:
I've added write error entry to CQ in case of failures. rxm_cq_reprocess_directed_recvs would not return a failure anymore.